packages/wrangler/src/api/startDevWorker/ConfigController.ts

```diff
@@ -1,67 +1,385 @@
+				"https://github.com/cloudflare/workers-sdk/issues/583."
+		);
+	}
+
+	// TODO(queues) support remote wrangler dev
+	const queues = extractBindingsOfType("queue", resolved.bindings);
+	if (
+		resolved.dev.remote &&
+		(queues?.length ||
+			resolved.triggers?.some((t) => t.type === "queue-consumer"))
+	) {
+		logger.warn(
+			"Queues are currently in Beta and are not supported in wrangler dev remote mode."
+		);
+	}
+	return resolved;
+}
 export class ConfigController extends Controller<ConfigControllerEventMap> {
-	latestInput?: Options;
-	latestConfig?: Options;
+	latestInput?: StartDevWorkerInput;
+	latestConfig?: StartDevWorkerOptions;
+
+	#configWatcher?: ReturnType<typeof watch>;
 
-	public set(input: Options) {
-		this.#updateConfig(input);
+	async #ensureWatchingConfig(configPath: string | undefined) {
+		await this.#configWatcher?.close();
+		if (configPath) {
+			this.#configWatcher = watch(configPath, {
+				persistent: true,
+			}).on("change", async (_event) => {
+				logger.log(`${path.basename(configPath)} changed...`);
+				assert(
+					this.latestInput,
+					"Cannot be watching config without having first set an input"
+				);
+				void this.#updateConfig(this.latestInput);
+			});
+		}
+	}
+	public async set(input: StartDevWorkerInput) {
+		return await this.#updateConfig(input);
 	}
-	public patch(input: Partial<Options>) {
+	public async patch(input: Partial<StartDevWorkerInput>) {
 		assert(
 			this.latestInput,
 			"Cannot call updateConfig without previously calling setConfig"
 		);
 
-		const config: Options = {
+		const config: StartDevWorkerInput = {
 			...this.latestInput,
 			...input,
 		};
 
-		this.#updateConfig(config);
+		return await this.#updateConfig(config);
 	}
 
-	#updateConfig(input: Options) {
-		const directory = input.directory;
-
-		this.latestConfig = {
-			directory,
-			build: {
-				moduleRules: [],
-				additionalModules: [],
-				define: {},
-				format: "modules",
-				moduleRoot: directory, // TODO: this default needs to come from getEntry() once readConfig has been moved into ConfigController
-				...input.build,
-			},
-			...input,
-		};
+	async #updateConfig(input: StartDevWorkerInput) {
 		this.latestInput = input;
-		this.emitConfigUpdateEvent(this.latestConfig);
+
+		const fileConfig = readConfig(input.config, {
+			env: input.env,
+			"dispatch-namespace": undefined,
+			"legacy-env": !input.legacy?.enableServiceEnvironments ?? true,
+			remote: input.dev?.remote,
+			upstreamProtocol:
+				input.dev?.origin?.secure === undefined
+					? undefined
+					: input.dev?.origin?.secure
+						? "https"
+						: "http",
+			localProtocol:
+				input.dev?.server?.secure === undefined
+					? undefined
+					: input.dev?.server?.secure
+						? "https"
+						: "http",
+		});
+		void this.#ensureWatchingConfig(fileConfig.configPath);
+
+		const resolvedConfig = await resolveConfig(fileConfig, input);
+		this.latestConfig = resolvedConfig;
+		this.emitConfigUpdateEvent(resolvedConfig);
```

Comment:
Need to avoid a race condition here where #updateConfig was called again before resolveConfig resolves

If we can't rely on reference identity of latestInput/fileConfig, then we should use an AbortController

Reply:
https://github.com/cloudflare/workers-sdk/pull/6161/commits/c80948bce8acc14dd4347762c2b0fb5bf615798b